backend/src/modules/url/adapter/googleapis.adapter.ts

```diff
@@ -0,0 +1,72 @@
+import { Injectable, Logger } from '@nestjs/common';
+import axios, { AxiosResponse } from 'axios';
+import { ApiConfigService } from '../../../shared/services/api-config.service';
+import { GoogleSafeBrowsingResponse } from 'src/common/types';
+
+@Injectable()
+export class GoogleApisAdapter {
+  private readonly logger = new Logger(GoogleApisAdapter.name);
+
+  constructor(private readonly apiConfigService: ApiConfigService) {}
+
+  async findUrlScore(url: string): Promise<number> {
+    const { clientId, clientVersion, safeBrowsingUrl } =
+      this.apiConfigService.googleApiConfig;
+
+    try {
+      const payload = {
+        client: { clientId, clientVersion },
+        threatInfo: {
+          threatTypes: [
+            'MALWARE', 
+            'SOCIAL_ENGINEERING',
+            'UNWANTED_SOFTWARE'
+          ],
+          platformTypes: [
+            'ANY_PLATFORM'
+          ],
+          threatEntryTypes: [
+            'URL'
+          ],
+          threatEntries: [
+            { url },
+          ],
+        },
+      };
+
+      const response: AxiosResponse<GoogleSafeBrowsingResponse> =
+        await axios.post(safeBrowsingUrl, payload);
+
+      let score = 1;
+      let threatType: string;
+
+      if (response.data.matches && response.data.matches.length > 0) {
+        threatType = response.data.matches[0].threatType;

```

Comment:
Đọc cái gọi là "Early returning pattern", ở đây chỗ này cần viết lại là


```suggestion
      const matches = response.data.matches;
      
      if (_.isEmpty(matches)) {
        return 1;
      }
      
      const threatType = matches[0].threatType;
      .....
```

Để ý em viết `response.data.matches` lặp lại tới 3 lần, nói chung 1 biểu thức dài dài mà lặp lại như vậy 2 lần thôi là nên đặt tên biến rồi